### DISCUSSION TITLE
Which version given PR was released in?

### DISCUSSION THREAD

**Discussion by findepi** [apache/datafusion#12999] [Ideas]
When looking at specific changes in DataFusion, sometimes one ends up wondering which version given PR was released in.

I cannot find a good way to see this in Github UI.

#### Proposed solution

Associated merged PRs with next release milestone, automatically.

**Comment by shanesveller** [comment]
FWIW this is a fairly common GitHub UI concern IME, and there are a few ways to answer this yourself even if the project doesn't make any procedural changes.

When viewing a given PR, if you identify its merged commit and visit the URL for that individual commit, it will describe which Git tag(s), if any, contain the commit in question. Since the Datafusion project reliably tags their releases, this should indicate which release(s) have included that change, assuming the code in question didn't evolve further since that PR.

Here are some "worked examples" of a few arbitrarily chosen PRs:

# Via GitHub UI

## 12986

https://github.com/apache/datafusion/pull/12986

> `@alamb` alamb merged commit [0011f45](https://github.com/apache/datafusion/commit/0011f450ada49fc3b724887eda14a788ec582f53) into apache:branch-42 [Oct 17, 2024](https://github.com/apache/datafusion/pull/12986#event-14699346008)

<details><summary>Screenshot of tag display for commit 0011f45</summary>
<img width="585" alt="image" src="https://github.com/user-attachments/assets/a50a75c2-40da-43ee-8aaf-354ca5d0130d">
</details>

This change (predictably) was first available in 42.1.0 RC1.

## 10000

https://github.com/apache/datafusion/pull/10000

> `@alamb` alamb merged commit [78f8ef1](https://github.com/apache/datafusion/commit/78f8ef16b89ffa38f80e8c7e3f21948602901787) into apache:main [Apr 8, 2024](https://github.com/apache/datafusion/pull/10000#event-12399335984)

<details><summary>Screenshot of tag display for commit 78f8ef1</summary>
<img width="916" alt="image" src="https://github.com/user-attachments/assets/4596e9ce-82bf-4973-b91a-3c3938971c13">
</details>

This change was first available in 38.0.0 RC1.

## 8000

https://github.com/apache/datafusion/pull/8000

> `@alamb` alamb merged commit [747cb50](https://github.com/apache/datafusion/commit/747cb502c47dec4c78f03f0d0ef2aabb447f8cfa) into apache:main [Oct 31, 2023](https://github.com/apache/datafusion/pull/8000#event-10821661034)

<details><summary>Screenshot of tag display for commit 747cb50</summary>
<img width="624" alt="image" src="https://github.com/user-attachments/assets/4088bb14-947c-4c4c-81ba-02554f29aec4">
</details>

This change was first available in 33.0.0 RC1.

# CLI

This information can also be determined using vanilla Git CLI commands if you have a local clone of Datafusion.

```bash
$ git tag --contains 0011f45
42.1.0-rc1
$ git tag --contains 78f8ef1 --sort=version:refname
38.0.0      38.0.0-rc1  39.0.0      39.0.0-rc1  40.0.0      40.0.0-rc1  41.0.0      41.0.0-rc1  42.0.0      42.0.0-rc1  42.1.0-rc1
$ git tag --contains 747cb50 --sort=version:refname
33.0.0      34.0.0      34.0.0-rc3  36.0.0      36.0.1-rc1  37.0.0-rc2  37.1.0-rc2  39.0.0      40.0.0-rc1  42.0.0
33.0.0-rc1  34.0.0-rc1  35.0.0      36.0.0-rc1  37.0.0      37.1.0      38.0.0      39.0.0-rc1  41.0.0      42.0.0-rc1
33.0.0-rc2  34.0.0-rc2  35.0.0-rc1  36.0.1      37.0.0-rc1  37.1.0-rc1  38.0.0-rc1  40.0.0      41.0.0-rc1  42.1.0-rc1
```

I assume there is some way to query the GitHub API, perhaps via their `gh` CLI, to obtain the mapping from merged PR number to merged commit, but I haven't a good reason to work that out so far.

**Comment by alamb** [comment]
> Associated merged PRs with next release milestone, automatically.

This sounds like a good idea to me @findepi  -- do you know how to configure / set this up?

**Comment by findepi** [comment]
@shanesveller thanks for explaining this. I know this, but this will surely be helpful for others.
My point is -- this is quite a bit of work, whereas the information could be readily available and easily accessible .

> This sounds like a good idea to me @findepi -- do you know how to configure / set this up?

@alamb i would be tempted to copy a solution form some other OSS project that has it.
For example Trino does this here: https://github.com/trinodb/trino/blob/e2a01389221326a9468fddd6b318206fe2acf0ea/.github/workflows/milestone.yml
The necessary ingredient is being able to derive the version string of the next release from the source code.
I don't know how we _should_ do this, but we _might_ do this by the following: change release process so that the version is incremented right after the release, not before. 
Some other way is to always assign PRs to `next-release` milestone and rename that milestone during the release process.

@alamb @andygrove thoughts?

**Reply by alamb** [reply]
In my opinion, adding version information to issues / PRs is a value to users so if we can figure out how to do it, we should

However, given I think maintainer bandwidth is the most limited resource we have, any solution we come up with should require little / no additional effort during code review.

> @alamb i would be tempted to copy a solution form some other OSS project that has it.
> For example Trino does this here: https://github.com/trinodb/trino/blob/e2a01389221326a9468fddd6b318206fe2acf0ea/.github/workflows/milestone.yml

I think it is a great idea. 

One idea I had was what if we added github labels retroactively **after** the release?  For example, we could write a script that ran over all commits and added a label for the first commit it appears in. There is some version of such a script (used for components in arrow-rs: https://github.com/apache/arrow-rs/blob/master/dev/release/label_issues.py.

**Reply by findepi** [reply]
The proposed solution would work, except that recently merged PRs would lack information we want to provide. 

Let's maybe go with `next-release` milestone, which is renamed during release. WDYT?

**Reply by alamb** [reply]
> The proposed solution would work, except that recently merged PRs would lack information we want to provide.
> 
> Let's maybe go with `next-release` milestone, which is renamed during release. WDYT?

Seems like a reasonable approach to me -- if you can document it we can give it a try for this the 43 release maybe
